daphne_worker/src/durable/garbage_collector.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2022 Cloudflare, Inc. All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+
+use crate::{durable, int_err, now};
+use rand::prelude::*;
+use worker::*;
+
+pub(crate) const DURABLE_GARBAGE_COLLECTOR_PUT: &str = "/internal/do/garbage_collector/put";
+
+/// Durable Object (DO) for keeping track of all persistent DO storage.
+#[durable_object]
+pub struct GarbageCollector {
+    #[allow(dead_code)]
+    state: State,
+    env: Env,
+}
+
+#[durable_object]
+impl DurableObject for GarbageCollector {
+    fn new(state: State, env: Env) -> Self {
+        Self { state, env }
+    }
+
+    async fn fetch(&mut self, mut req: Request) -> Result<Response> {
+        let mut rng = thread_rng();
+        match (req.path().as_ref(), req.method()) {
+            // Schedule a durable object (DO) instance for deletion.
+            (DURABLE_GARBAGE_COLLECTOR_PUT, Method::Post) => {
+                let durable_ref: durable::DurableReference = req.json().await?;
+                match durable_ref.binding.as_ref() {
+                    durable::BINDING_DAP_REPORT_STORE
+                    | durable::BINDING_DAP_AGGREGATE_STORE
+                    | durable::BINDING_DAP_LEADER_AGG_JOB_QUEUE
+                    | durable::BINDING_DAP_LEADER_COL_JOB_QUEUE
+                    | durable::BINDING_DAP_HELPER_STATE_STORE => (),
+                    s => {
+                        return Err(int_err(format!(
+                            "GarbageCollector: unrecognized binding: {}",
+                            s
+                        )))
+                    }
+                };
+
+                // Reference handle encodes the current time and a random nonce in order to deal
+                // with collisions. This time goes first in order to ensure we can delete the
+                // oldest objects first.
```

Comment:
Reviewer: Please sanity check that this encoding actually results in things getting handled in order by (binding, timestamp).

Reply:
this depends on whether the function returns elements sorted, and whether the map returned is sorted too. An extra countermeasure is to sort-again or check if it's sorted, but it may be too much overhead.